### PR TITLE
preparing bootstrap onboarding request

### PR DESCRIPTION
**Pull request by alknopfler**
WIP preparing bootstrap onboarding request

### CODE REVIEW

File: sztp-agent/cmd/daemon.go [hunk 1]
```diff
@@ -34,11 +34,11 @@ func NewDaemonCommand() *cobra.Command {
 	flags := cmd.Flags()
 	//TODO this options should be retrieved automatically instead of requests in the agent
 	//Opened discussion to define the procedure: https://github.com/opiproject/sztp/issues/2
-	flags.StringVar(&serialNumber, "serial-number", "", "Device's serial number")
-	flags.StringVar(&devicePassword, "device-password", "", "Device's password")
-	flags.StringVar(&devicePrivateKey, "device-private-key", "", "Device's private key")
-	flags.StringVar(&deviceEndEntityCert, "device-end-entity-cert", "", "Device's End Entity cert")
-	flags.StringVar(&bootstrapTrustAnchorCert, "bootstrap-trust-anchor-cert", "", "Bootstrap server trust anchor Cert")
+	flags.StringVar(&serialNumber, "serial-number", "my-serial", "Device's serial number")
```
Comment (alknopfler) [inline sztp-agent/cmd/daemon.go:37]: @glimchb by default the param use my-serial and so on... Let's keep in mind when we know how get it from Vendor

File: sztp-agent/cmd/daemon.go [hunk 2]
```diff
@@ -34,11 +34,11 @@ func NewDaemonCommand() *cobra.Command {
...
+	flags.StringVar(&devicePassword, "device-password", "mysecret", "Device's password")
+	flags.StringVar(&devicePrivateKey, "device-private-key", "client/end-entity/private_key.pem", "Device's private key")
+	flags.StringVar(&deviceEndEntityCert, "device-end-entity-cert", "client/end-entity/my_cert.pem", "Device's End Entity cert")
+	flags.StringVar(&bootstrapTrustAnchorCert, "bootstrap-trust-anchor-cert", "/opi.pem", "Bootstrap server trust anchor Cert")
```
Comment (glimchb) [inline sztp-agent/cmd/daemon.go:41]: the keys and certificates added here https://github.com/opiproject/sztp/pull/49

File: sztp-agent/cmd/daemon.go [hunk 3]
```diff
@@ -34,11 +34,11 @@ func NewDaemonCommand() *cobra.Command {
 	flags := cmd.Flags()
 	//TODO this options should be retrieved automatically instead of requests in the agent
 	//Opened discussion to define the procedure: https://github.com/opiproject/sztp/issues/2
-	flags.StringVar(&serialNumber, "serial-number", "", "Device's serial number")
-	flags.StringVar(&devicePassword, "device-password", "", "Device's password")
-	flags.StringVar(&devicePrivateKey, "device-private-key", "", "Device's private key")
-	flags.StringVar(&deviceEndEntityCert, "device-end-entity-cert", "", "Device's End Entity cert")
-	flags.StringVar(&bootstrapTrustAnchorCert, "bootstrap-trust-anchor-cert", "", "Bootstrap server trust anchor Cert")
```
Comment (glimchb) [inline sztp-agent/cmd/daemon.go:41]: should be `my-serial-number` , right ?

File: sztp-agent/pkg/secureAgent/agent.go [hunk 4]
```diff
@@ -7,6 +7,27 @@ Copyright (C) 2022 Red Hat.
 
 package secureAgent
 
+const (
+	CONTENT_TYPE_YANG = "Content-Type:application/yang-data+json"
```
Comment (glimchb) [inline sztp-agent/pkg/secureAgent/agent.go:11]: should be `"application/yang-data+json"` not `"Content-Type:application/yang-data+json"`

File: sztp-agent/pkg/secureAgent/agent.go [hunk 5]
```diff
@@ -7,6 +7,27 @@ Copyright (C) 2022 Red Hat.
...
+	OS_RELEASE_FILE   = "/etc/os-release"
+)
+
+type InputJSON struct {
+	IetfSztpBootstrapServerInput struct {
+		HwModel             string        `json:"hw-model"`
+		OsName              string        `json:"os-name"`
+		OsVersion           string        `json:"os-version"`
+		SignedDataPreferred []interface{} `json:"signed-data-preferred"`
```
Comment (glimchb) [inline sztp-agent/pkg/secureAgent/agent.go:20]: `SignedDataPreferred` is wrong here, go error
```
{
  "ietf-restconf:errors": {
    "error": [
      {
        "error-type": "protocol",
        "error-tag": "malformed-message",
        "error-message": "Unable to parse \"input\" document: Doesn't match schema: {/ietf-sztp-bootstrap-server:input/signed-data-preferred} expected empty value"
      }
    ]
  }
}
```

File: sztp-agent/pkg/secureAgent/utils.go [hunk 6]
```diff
@@ -20,7 +29,79 @@ func linesInFileContains(file string, substr string) string {
...
+			HwModel             string        `json:"hw-model"`
+			OsName              string        `json:"os-name"`
+			OsVersion           string        `json:"os-version"`
+			SignedDataPreferred []interface{} `json:"signed-data-preferred"`
+			Nonce               string        `json:"nonce"`
+		}{
+			HwModel:             "hardwared-model-TBD",
+			OsName:              osName,
+			OsVersion:           osVersion,
+			SignedDataPreferred: nil,
```
Comment (glimchb) [inline sztp-agent/pkg/secureAgent/utils.go:97]: `SignedDataPreferred` is wrong here, got this error
```
{
  "ietf-restconf:errors": {
    "error": [
      {
        "error-type": "protocol",
        "error-tag": "malformed-message",
        "error-message": "Unable to parse \"input\" document: Doesn't match schema: {/ietf-sztp-bootstrap-server:input/signed-data-preferred} expected empty value"
      }
    ]
  }
}
```

File: sztp-agent/pkg/secureAgent/agent.go (hunk 4 above)
Comment (glimchb) [inline sztp-agent/pkg/secureAgent/agent.go:11]:
```suggestion
	CONTENT_TYPE_YANG = "application/yang-data+json"
```

File: sztp-agent/cmd/daemon.go (hunk 1 above)
Comment (glimchb) [inline sztp-agent/cmd/daemon.go:37]:
```suggestion
	flags.StringVar(&serialNumber, "serial-number", "my-serial-number", "Device's serial number")
```

File: sztp-agent/pkg/secureAgent/utils.go [hunk 7]
```diff
@@ -20,7 +29,79 @@ func linesInFileContains(file string, substr string) string {
...
+			},
+		},
+	}
+	log.Println(r)
+	time.Sleep(20 * time.Second)
+	res, err := client.Do(r)
+	defer res.Body.Close()
+
+	post := &BootstrapServerPostOutput{}
+	derr := json.NewDecoder(res.Body).Decode(post)
```
Comment (glimchb) [inline sztp-agent/pkg/secureAgent/utils.go:69]: I also used for debug to print the Body this code:
```
        b, err := io.ReadAll(res.Body)
        log.Println(string(b))
```

File: sztp-agent/pkg/secureAgent/daemon.go [hunk 8]
```diff
@@ -32,12 +32,18 @@ func (a *Agent) runDaemon() error {
...
 	}
-	log.Println("[INFO] Bootstrap URL retrieved successfully")
+	log.Println("[INFO] Bootstrap URL retrieved successfully.")
+	log.Println("[INFO] Starting the Request to get On-boarding Information.")
+	res, err := a.doTLSRequestToBootstrap()
+	if err != nil {
+		log.Println("[ERROR] ", err.Error())
+		return err
+	}
+	log.Println(res)
```
Comment (glimchb) [inline sztp-agent/pkg/secureAgent/daemon.go:47]: example of res json string is
```
{
  "ietf-sztp-bootstrap-server:output": {
    "conveyed-information": "MIIDfwYLKoZIhvcNAQkQASugggNuBIIDansKICAiaWV0Zi1zenRwLWNvbnZleWVkLWluZm86b25ib2FyZGluZy1pbmZvcm1hdGlvbiI6IHsKICAgICJib290LWltYWdlIjogewogICAgICAiZG93bmxvYWQtdXJpIjogWwogICAgICAgICJodHRwOi8vd2ViOjgwODIvdmFyL2xpYi9taXNjL215LWJvb3QtaW1hZ2UuaW1nIiwKICAgICAgICAiZnRwOi8vd2ViOjMwODIvdmFyL2xpYi9taXNjL215LWJvb3QtaW1hZ2UuaW1nIgogICAgICBdLAogICAgICAiaW1hZ2UtdmVyaWZpY2F0aW9uIjogWwogICAgICAgIHsKICAgICAgICAgICJoYXNoLWFsZ29yaXRobSI6ICJpZXRmLXN6dHAtY29udmV5ZWQtaW5mbzpzaGEtMjU2IiwKICAgICAgICAgICJoYXNoLXZhbHVlIjogIjdiOmNhOmU2OmFjOjIzOjA2OmQ4Ojc5OjA2OjhjOmFjOjAzOjgwOmUyOjE2OjQ0OjdlOjQwOjZhOjY1OmZhOmQ0OjY5OjYxOjZlOjA1OmNlOmY1Ojg3OmRjOjJiOjk3IgogICAgICAgIH0KICAgICAgXQogICAgfSwKICAgICJwcmUtY29uZmlndXJhdGlvbi1zY3JpcHQiOiAiSXlFdlltbHVMMkpoYzJnS1pXTm9ieUFpYVc1emFXUmxJSFJvWlNCd2NtVXRZMjl1Wm1sbmRYSmhkR2x2YmkxelkzSnBjSFF1TGk0aUNnPT0iLAogICAgImNvbmZpZ3VyYXRpb24taGFuZGxpbmciOiAibWVyZ2UiLAogICAgImNvbmZpZ3VyYXRpb24iOiAiUEhSdmNDQjRiV3h1Y3owaWFIUjBjSE02TDJWNFlXMXdiR1V1WTI5dEwyTnZibVpwWnlJK0NpQWdQR0Z1ZVMxNGJXd3RZMjl1ZEdWdWRDMXZhMkY1THo0S1BDOTBiM0ErQ2c9PSIsCiAgICAicG9zdC1jb25maWd1cmF0aW9uLXNjcmlwdCI6ICJJeUV2WW1sdUwySmhjMmdLWldOb2J5QWlhVzV6YVdSbElIUm9aU0J3YjNOMExXTnZibVpwWjNWeVlYUnBiMjR0YzJOeWFYQjBMaTR1SWdvPSIKICB9Cn0="
  }
}
```
now can base64 decode it like this
```
    rawDecodedText, err := base64.StdEncoding.DecodeString(res)
```

File: sztp-agent/pkg/secureAgent/utils.go (hunk 7 above)
Comment (alknopfler) [reply sztp-agent/pkg/secureAgent/utils.go:69]: I've got the response struct log, the same but in struct

**Comment by alknopfler** [comment]
solved some issues. From my side, ready to merge it 
image

**Review by glimchb — APPROVED** [review]
Excellent
```
opi-sztp-go-agent_1  | 2022/10/14 11:28:06 {{MIIDfwYLKoZIhvcNAQkQASugggNuBIIDansKICAiaWV0Zi1zenRwLWNvbnZleWVkLWluZm86b25ib2FyZGluZy1pbmZvcm1hdGlvbiI6IHsKICAgICJib290LWltYWdlIjogewogICAgICAiZG93bmxvYWQtdXJpIjogWwogICAgICAgICJodHRwOi8vd2ViOjgwODIvdmFyL2xpYi9taXNjL215LWJvb3QtaW1hZ2UuaW1nIiwKICAgICAgICAiZnRwOi8vd2ViOjMwODIvdmFyL2xpYi9taXNjL215LWJvb3QtaW1hZ2UuaW1nIgogICAgICBdLAogICAgICAiaW1hZ2UtdmVyaWZpY2F0aW9uIjogWwogICAgICAgIHsKICAgICAgICAgICJoYXNoLWFsZ29yaXRobSI6ICJpZXRmLXN6dHAtY29udmV5ZWQtaW5mbzpzaGEtMjU2IiwKICAgICAgICAgICJoYXNoLXZhbHVlIjogIjdiOmNhOmU2OmFjOjIzOjA2OmQ4Ojc5OjA2OjhjOmFjOjAzOjgwOmUyOjE2OjQ0OjdlOjQwOjZhOjY1OmZhOmQ0OjY5OjYxOjZlOjA1OmNlOmY1Ojg3OmRjOjJiOjk3IgogICAgICAgIH0KICAgICAgXQogICAgfSwKICAgICJwcmUtY29uZmlndXJhdGlvbi1zY3JpcHQiOiAiSXlFdlltbHVMMkpoYzJnS1pXTm9ieUFpYVc1emFXUmxJSFJvWlNCd2NtVXRZMjl1Wm1sbmRYSmhkR2x2YmkxelkzSnBjSFF1TGk0aUNnPT0iLAogICAgImNvbmZpZ3VyYXRpb24taGFuZGxpbmciOiAibWVyZ2UiLAogICAgImNvbmZpZ3VyYXRpb24iOiAiUEhSdmNDQjRiV3h1Y3owaWFIUjBjSE02TDJWNFlXMXdiR1V1WTI5dEwyTnZibVpwWnlJK0NpQWdQR0Z1ZVMxNGJXd3RZMjl1ZEdWdWRDMXZhMkY1THo0S1BDOTBiM0ErQ2c9PSIsCiAgICAicG9zdC1jb25maWd1cmF0aW9uLXNjcmlwdCI6ICJJeUV2WW1sdUwySmhjMmdLWldOb2J5QWlhVzV6YVdSbElIUm9aU0J3YjNOMExXTnZibVpwWjNWeVlYUnBiMjR0YzJOeWFYQjBMaTR1SWdvPSIKICB9Cn0=}}
opi-sztp-go-agent_1  | 2022/10/14 11:28:06 &{{0
        +��n�j{                                *�H��
opi-sztp-go-agent_1  |   "ietf-sztp-conveyed-info:onboarding-information": {
opi-sztp-go-agent_1  |     "boot-image": {
opi-sztp-go-agent_1  |       "download-uri": [
opi-sztp-go-agent_1  |         "http://web:8082/var/lib/misc/my-boot-image.img",
opi-sztp-go-agent_1  |         "ftp://web:3082/var/lib/misc/my-boot-image.img"
opi-sztp-go-agent_1  |       ],
opi-sztp-go-agent_1  |       "image-verification": [
opi-sztp-go-agent_1  |         {
opi-sztp-go-agent_1  |           "hash-algorithm": "ietf-sztp-conveyed-info:sha-256",
opi-sztp-go-agent_1  |           "hash-value": "7b:ca:e6:ac:23:06:d8:79:06:8c:ac:03:80:e2:16:44:7e:40:6a:65:fa:d4:69:61:6e:05:ce:f5:87:dc:2b:97"
opi-sztp-go-agent_1  |         }
opi-sztp-go-agent_1  |       ]
opi-sztp-go-agent_1  |     },
opi-sztp-go-agent_1  |     "pre-configuration-script": "IyEvYmluL2Jhc2gKZWNobyAiaW5zaWRlIHRoZSBwcmUtY29uZmlndXJhdGlvbi1zY3JpcHQuLi4iCg==",
opi-sztp-go-agent_1  |     "configuration-handling": "merge",
opi-sztp-go-agent_1  |     "configuration": "PHRvcCB4bWxucz0iaHR0cHM6L2V4YW1wbGUuY29tL2NvbmZpZyI+CiAgPGFueS14bWwtY29udGVudC1va2F5Lz4KPC90b3A+Cg==",
opi-sztp-go-agent_1  |     "post-configuration-script": "IyEvYmluL2Jhc2gKZWNobyAiaW5zaWRlIHRoZSBwb3N0LWNvbmZpZ3VyYXRpb24tc2NyaXB0Li4uIgo="
opi-sztp-go-agent_1  |   }
opi-sztp-go-agent_1  | }}}
sztp_opi-sztp-go-agent_1 exited with code 0
```

**Pull request merged**